### ISSUE TITLE
TransactionIsStillOpenException while trying to PUT a document

### ISSUE THREAD

**Issue by relgames** [baasbox/baasbox#960]
Making request PUT http://localhost:8080/document/checkins/e4883182-182f-4320-9fa3-24b1290ede39 with JSON content, but getting 500 Internal Server Error:

``` javascript
{
  "result": "error",
  "message": "com.baasbox.exception.TransactionIsStillOpenException : Controller left an open transaction. Database will be rollbacked",
  "stacktrace": [
    "com.baasbox.controllers.actions.filters.ConnectToDBFilter.call((ConnectToDBFilter.java:83)",
    "com.baasbox.controllers.actions.filters.UserCredentialWrapFilter.call((UserCredentialWrapFilter.java:74)",
    "play.core.j.JavaAction$$anon$3.apply((JavaAction.scala:91)",
    "play.core.j.JavaAction$$anon$3.apply((JavaAction.scala:90)",
    "play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply((FPromiseHelper.scala:82)",
    "play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply((FPromiseHelper.scala:82)",
    "scala.concurrent.Future$$anonfun$flatMap$1.apply((Future.scala:251)",
    "scala.concurrent.Future$$anonfun$flatMap$1.apply((Future.scala:249)",
    "scala.concurrent.impl.CallbackRunnable.run((Promise.scala:32)",
    "play.core.j.HttpExecutionContext$$anon$2.run((HttpExecutionContext.scala:37)",
    "akka.dispatch.TaskInvocation.run((AbstractDispatcher.scala:42)",
    "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec((AbstractDispatcher.scala:386)",
    "scala.concurrent.forkjoin.ForkJoinTask.doExec((ForkJoinTask.java:260)",
    "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask((ForkJoinPool.java:1339)",
    "scala.concurrent.forkjoin.ForkJoinPool.runWorker((ForkJoinPool.java:1979)",
    "scala.concurrent.forkjoin.ForkJoinWorkerThread.run((ForkJoinWorkerThread.java:107)"
  ],
  "full_stacktrace": "com.baasbox.exception.TransactionIsStillOpenException: Controller left an open transaction. Database will be rollbacked\n\tat com.baasbox.controllers.actions.filters.ConnectToDBFilter.call(ConnectToDBFilter.java:83)\n\tat com.baasbox.controllers.actions.filters.UserCredentialWrapFilter.call(UserCredentialWrapFilter.java:74)\n\tat play.core.j.JavaAction$$anon$3.apply(JavaAction.scala:91)\n\tat play.core.j.JavaAction$$anon$3.apply(JavaAction.scala:90)\n\tat play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply(FPromiseHelper.scala:82)\n\tat play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply(FPromiseHelper.scala:82)\n\tat scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)\n\tat scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249)\n\tat scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)\n\tat play.core.j.HttpExecutionContext$$anon$2.run(HttpExecutionContext.scala:37)\n\tat akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:42)\n\tat akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:386)\n\tat scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)\n\tat scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)\n\tat scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)\n\tat scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\n",
  "resource": "/document/checkins/e4883182-182f-4320-9fa3-24b1290ede39",
  "method": "PUT",
  "request_header": {
    "X-BAASBOX-APPCODE": [
      "app"
    ],
    "Content-Length": [
      "904"
    ],
    "X-BB-SESSION": [
      "xxx"
    ],
    "Content-Type": [
      "application/json"
    ],
    "User-Agent": [
      "akka-http/2.4.10"
    ],
    "Host": [
      "localhost:8080"
    ]
  },
  "API_version": "0.9.5",
  "http_code": 500
}
```

**Comment by giastfader** [comment]
Hi, something happened just before this.
You should check the application log and see if any exception has been logged there.

**Comment by relgames** [comment]
Hi, it happened again. I do not see any errors before that:

``` json
{
  "result": "error",
  "message": "com.baasbox.exception.TransactionIsStillOpenException : Controller left an open transaction. Database will be rollbacked",
  "stacktrace": [
    "com.baasbox.controllers.actions.filters.ConnectToDBFilter.call((ConnectToDBFilter.java:83)",
    "com.baasbox.controllers.actions.filters.UserCredentialWrapFilter.call((UserCredentialWrapFilter.java:74)",
    "play.core.j.JavaAction$$anon$3.apply((JavaAction.scala:91)",
    "play.core.j.JavaAction$$anon$3.apply((JavaAction.scala:90)",
    "play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply((FPromiseHelper.scala:82)",
    "play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply((FPromiseHelper.scala:82)",
    "scala.concurrent.Future$$anonfun$flatMap$1.apply((Future.scala:251)",
    "scala.concurrent.Future$$anonfun$flatMap$1.apply((Future.scala:249)",
    "scala.concurrent.impl.CallbackRunnable.run((Promise.scala:32)",
    "play.core.j.HttpExecutionContext$$anon$2.run((HttpExecutionContext.scala:37)",
    "akka.dispatch.TaskInvocation.run((AbstractDispatcher.scala:42)",
    "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec((AbstractDispatcher.scala:386)",
    "scala.concurrent.forkjoin.ForkJoinTask.doExec((ForkJoinTask.java:260)",
    "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask((ForkJoinPool.java:1339)",
    "scala.concurrent.forkjoin.ForkJoinPool.runWorker((ForkJoinPool.java:1979)",
    "scala.concurrent.forkjoin.ForkJoinWorkerThread.run((ForkJoinWorkerThread.java:107)"
  ],
  "full_stacktrace": "com.baasbox.exception.TransactionIsStillOpenException: Controller left an open transaction. Database will be rollbacked\n\tat com.baasbox.controllers.actions.filters.ConnectToDBFilter.call(ConnectToDBFilter.java:83)\n\tat com.baasbox.controllers.actions.filters.UserCredentialWrapFilter.call(UserCredentialWrapFilter.java:74)\n\tat play.core.j.JavaAction$$anon$3.apply(JavaAction.scala:91)\n\tat play.core.j.JavaAction$$anon$3.apply(JavaAction.scala:90)\n\tat play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply(FPromiseHelper.scala:82)\n\tat play.core.j.FPromiseHelper$$anonfun$flatMap$1.apply(FPromiseHelper.scala:82)\n\tat scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)\n\tat scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249)\n\tat scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)\n\tat play.core.j.HttpExecutionContext$$anon$2.run(HttpExecutionContext.scala:37)\n\tat akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:42)\n\tat akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:386)\n\tat scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)\n\tat scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)\n\tat scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)\n\tat scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\n",
  "resource": "/document/checkins/5f3a1214-8377-4407-96b7-e29b262d7aa6",
  "method": "PUT",
  "request_header": {
    "X-BAASBOX-APPCODE": [
      "app"
    ],
    "Content-Length": [
      "652"
    ],
    "X-BB-SESSION": [
      "xx"
    ],
    "Content-Type": [
      "application/json"
    ],
    "User-Agent": [
      "akka-http/2.4.10"
    ],
    "Host": [
      "localhost:8080"
    ]
  },
  "API_version": "0.9.5",
  "http_code": 500
}
```

**Comment by giastfader** [comment]
This is strange, something must happen just before.
Try to start BaasBox enabling the verbose log:

```
./start  -Dlogger.resource=application-logger-debug.xml
```

**Comment by relgames** [comment]
Enabled logging. The only error I see here is `Cound not retrieve the DB connection within this thread: ODatabaseException: Database instance is not set in current thread. Assure to set it with: ODatabaseRecordThreadLocal.INSTANCE.set(db);`

Here is the log: http://pastebin.com/raw/iCtk8tZh
(log lines are not ordered by timestamp by some reason...)

**Comment by relgames** [comment]
Also, I see errors like:

```
com.orientechnologies.common.concur.lock.OLockException: Error on releasing a non acquired lock by the requester 'Thread[play-akka.actor.default-dispatcher-22,5,main]' against the resource: '#11:159'
```

**Comment by relgames** [comment]
I even tried to limit number of concurrent BaasBox API calls to 1, but still the same error happened.
Logs are full of OLockException and ODatabaseException.

**Comment by giastfader** [comment]
Those "exceptions" are actually warnings of the ODB engine. You can ignore them.
The issue is somewhere else.
Can you share the db with me?
In case, please send me a private email.

**Comment by relgames** [comment]
It happens randomly, I'm still trying to reproduce it.

**Comment by relgames** [comment]
Finally, I was able to reproduce it. The error message is really cryptic and driving away from the real cause.
The real cause is the user not having write permissions for the document in PUT.
I guess it must be changed to something like "Permission denied".

**Comment by giastfader** [comment]
You are using v.0.9.5
Less cryptic error has already been implemented in 1.0.0-\* versions
